### ISSUE TITLE
Signals not passing the object correctly when using builder

### ISSUE THREAD

**Issue by kadinparker** [StefanSalewski/gintro#45]
Hello, when using builder to get gtk objects from a glade file, the signals don't send the object. 
An example:

```nim
import gintro/[gtk, glib, gobject, gio, gtksource]

proc entryActivated (entry: Entry) =
  echo repr entry

proc appActivate(app: Application) =
  var builder = newBuilder()
  discard addFromFile(builder, "UI")
  let window = getApplicationWindow(builder, "window")
  window.setApplication(app)
  let entry = getEntry(builder, "inputbox")
  entry.connect("activate", entryActivated)
  showAll(window)

proc main =
  let app = newApplication("org.gtk.example")
  connect(app, "activate", appActivate)
  discard run(app)

main()
```
When the entry is activated, `nil` is output.
However, when you pass the entry as extra data, as follows
```nim
import gintro/[gtk, glib, gobject, gio, gtksource]

proc entryActivated (entry: Entry, entry1: Entry) =
  echo repr entry1

proc appActivate(app: Application) =
  var builder = newBuilder()
  discard addFromFile(builder, "UI")
  let window = getApplicationWindow(builder, "window")
  window.setApplication(app)
  let entry = getEntry(builder, "inputbox")
  entry.connect("activate", entryActivated, entry)
  showAll(window)

proc main =
  let app = newApplication("org.gtk.example")
  connect(app, "activate", appActivate)
  discard run(app)

main()
```
This passes the object correctly, and outputs 
```ref 0x7f393d6160a8 --> [impl = ref 0x558a52bc2ac0 --> [gTypeInstance = [gClass = ref 0x558a52e9a000 --> [gType = 94052581731728]],
refCount = 8,
qdata = ref 0x558a52e89e20 --> []]]
```
UI file is as follows
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!-- Generated with glade 3.22.1 -->
<interface>
  <requires lib="gtk+" version="3.20"/>
  <object class="GtkApplicationWindow" id="window">
    <property name="can_focus">False</property>
    <child>
      <placeholder/>
    </child>
    <child>
      <object class="GtkBox">
        <property name="visible">True</property>
        <property name="can_focus">False</property>
        <property name="orientation">vertical</property>
        <child>
          <object class="GtkTextView" id="messages">
            <property name="visible">True</property>
            <property name="can_focus">True</property>
            <property name="margin_bottom">5</property>
            <property name="vexpand">True</property>
            <property name="editable">False</property>
            <property name="cursor_visible">False</property>
          </object>
          <packing>
            <property name="expand">True</property>
            <property name="fill">True</property>
            <property name="position">0</property>
          </packing>
        </child>
        <child>
          <object class="GtkEntry" id="inputbox">
            <property name="visible">True</property>
            <property name="can_focus">True</property>
            <property name="valign">end</property>
            <property name="vexpand">False</property>
            <property name="placeholder_text" translatable="yes">Enter a message here</property>
            <signal name="activate" handler="entryActivated" swapped="no"/>
          </object>
          <packing>
            <property name="expand">False</property>
            <property name="fill">False</property>
            <property name="pack_type">end</property>
            <property name="position">1</property>
          </packing>
        </child>
      </object>
    </child>
  </object>
</interface>
```

Thanks

**Comment by StefanSalewski** [comment]
Thanks for reporting.

Currently I can not really remember your name, so I wonder how you have found function getEntry(builder) at all. I have added these functions recently for someone who wants to use gtk builder, but I think I have not provided an example in the tutorial.

I do not know much about gtk builder unfortunately, I had once tried it, but found it very hard to use -- not only from  Nim, but generally.

When I remember correctly signals can be connected from inside the glade files by specifing the names of the callback functions. Of course that works only when the function symbols can be found. I guess we have to try and investigate that. Maybe it will make a difference if we use a separate textual ui file or create a resource file from it and link it with the executable. I think I did the last for my NEd editor, it was a bit hard, but I can not remember details.

For your issue: Yes, seems to be obvious when we compare proc getEntry(builder) and proc newEntry() in  ~/.nimble/pkgs/gintro-0.4.17/gintro/gtk.nim. The whole g_object_set_qdata() stuff is missing. I think I forgot that because the widget is owned and keept alive by builder. But without it we can not get back the Nim proxy object from the low level gtk widget.

Will try to fix that soon ...

**Comment by StefanSalewski** [comment]
Do you use current Nim devel compiler or still old 0.19.4?

Problem is, my local gintro contains some fixes for latest compiler already...

I just did a first try:

```nim
proc getEntry*(builder: Builder; name: string): Entry =
  let gt = g_type_from_name("GtkEntry")
  assert(gt != 0)
  let gobj = gtk_builder_get_object(cast[ptr Builder00](builder.impl), name)
  if g_object_get_qdata(gobj, Quark) != nil:
    result = cast[type(result)](g_object_get_qdata(gobj, Quark))
    assert(result.impl == gobj)
  else:
    new(result, finalizeGObject)
    result.impl = gobj
    GC_ref(result)
    discard g_object_ref_sink(result.impl)
    g_object_add_toggle_ref(result.impl, toggleNotify, addr(result[]))
    g_object_unref(result.impl)
    assert(g_object_get_qdata(result.impl, Quark) == nil)
    g_object_set_qdata(result.impl, Quark, addr(result[]))
```

That makes your provided example work. But I still have to investigate how memory management works in detail for builder, API description is not that clear

https://developer.gnome.org/gtk3/stable/GtkBuilder.html

For example I do not yet know if objects get from builder are floating and which ref count they have. And one problem may result from Nim's finalizers, see my recent forum post

https://forum.nim-lang.org/t/4780

But I don't know is that is a problem or not. Will investigate.

**Comment by StefanSalewski** [comment]
Should be fixed with latest v0.4.18

I hope it will compile with old Nim 0.19.4 still.

For me your example works now:

```nim
import gintro/[gtk, glib, gobject, gio]

proc entryActivated (entry: Entry) =
  echo "Input: ", entry.text

proc appActivate(app: Application) =
  var builder = newBuilder()
  discard addFromFile(builder, "UI")
  let window = getApplicationWindow(builder, "window")
  window.setApplication(app)
  let entry = getEntry(builder, "inputbox")
  entry.connect("activate", entryActivated)
  showAll(window)

proc main =
  let app = newApplication("org.gtk.example")
  connect(app, "activate", appActivate)
  discard run(app)

main()
```

Connecting signals from within builder is still untested.

**Comment by StefanSalewski** [comment]
Working example is provided now:

https://github.com/StefanSalewski/gintro#gtk-builderuser-interfaces-created-with-the-glade-tool
